src/server/editorServices.ts

```diff
@@ -1285,11 +1304,65 @@ namespace ts.server {
             return configFileErrors;
         }
 
-        createInferredProjectWithRootFileIfNecessary(root: ScriptInfo) {
-            const useExistingProject = this.useSingleInferredProject && this.inferredProjects.length;
-            const project = useExistingProject
-                ? this.inferredProjects[0]
-                : new InferredProject(this, this.documentRegistry, this.compilerOptionsForInferredProjects);
+        private getOrCreateInferredProjectForProjectRootPathIfEnabled(root: ScriptInfo, projectRootPath: string | undefined): InferredProject | undefined {
+            if (!this.useInferredProjectPerProjectRoot) {
+                return undefined;
+            }
+
+            if (projectRootPath) {
+                // if we have an explicit project root path, find (or create) the matching inferred project.
+                for (const project of this.inferredProjects) {
+                    if (project.projectRootPath === projectRootPath) {
+                        return project;
+                    }
+                }
+                return this.createInferredProject(/*isSingleInferredProject*/ false, projectRootPath);
+            }
+
+            // we don't have an explicit root path, so we should try to find an inferred project that best matches the file.

```

Comment:
Define best -- it looks like this is finding the project containing the file that is the most deeply nested, so given if we're looking for a project for `/a/b/c.ts` and we have a choice of projects at `/a`, `/a/b`, and `/a/b/x`, we'll choose `/a/b`.